### DISCUSSION TITLE
Типы Bundle для выгрузки результатов лаб исследований

### DISCUSSION THREAD

**Discussion by ai-fhir** [fhir-ru/core#110] [Лабораторная РГ]
Мы рассматриваем два возможных варианта:

* collection - A collection (type = "collection") consists of a series of 0 or more entries. No particular use with respect to the FHIR specification is associated with this Bundle. Each entry element SHALL contain a resource.
* document - A document Bundle (type = "document") consists of a series of entries, the first of which is a Composition. Each entry element SHALL contain a resource. See Documents for further information.

## Collection

Collection это просто какое-то объединение ресурсов. Чтобы понять что внутри collection лежит именно протокол лабораторного исследования нужно попытаться найти среди ресурсов ресурс Task с определенным типом. В нашем случае это может быть не критично, т.к. предполагается API, который будет принимать только лабораторные протоколы (?). 

## Document

Document - FHIR resources can be used to build documents that represent a composition: a coherent set of information that is a statement of healthcare information, including clinical observations and services. A document is an immutable set of resources with a fixed presentation that is authored and/or attested by humans, organizations and devices.

Document предполагает, что у нас всегда есть первый ресурс с типом Composition, который указывает, что именно это за документ. Часть смысловой нагрузки в нашем случае должна переехать (продублироваться) из Task в Composition. Кажется, что использование Document более универсально, если мы строим сеть обмена разными документами. Document вводить формальные правила подписи документа и включения стилей для человекочитаемого отображения. При этом Document иммутабелен, в случае обнаружения ошибки мы не можем просто изменить документ, мы обязаны создать новый и сказать, что он заменяет старый; такая логика позволяет избегать проблему, когда врач или СППР сделали вывод на основе старой версии данных, а после обновления их выводы стали ошибочны.

## +/-

Composition:
`+`
* Простой, не несет доп требований

`-`
* Переизобретаем часть логики, которая уже заложена в Bundle с типом document
* Мы можем наложить ограничения на структуру нашего Bundle, но не можем наложить ограничения на иммутабельность.

Document:
`+`
* Задает общий вид всех документов (не надо вводить правило про использование Task)
* Задает правила подписи и отображения документов

`-`
* Иммутабельные документы сложнее, трубуют больше затрат на разработку

**Comment by JustAsWalker** [comment]
Не совсем понимаю, что подразумевается под "выгрузкой результатов лаб исследований" - это обмен протоколами лабораторного исследования или возможна передача отдельных результатов? Если обмен протоколами, то это выглядит как Bundle типа document.

В описании не упоминается ресурс [DiagnosticReport](https://www.hl7.org/fhir/diagnosticreport.html), который может описать протокол лабораторного исследования и через атрибут result ссылаться на отдельные результаты в виде Observation. Именно по нему, а точнее по его типизации, можно понять, что передается лабораторное исследование.

Кроме того, упоминается Task - для чего он используется в данном случае? Так описывается направление на лабораторное исследование? Если да, то разве это не ресурс [ServiceRequest](https://www.hl7.org/fhir/servicerequest.html)?

> врач или СППР сделали вывод на основе старой версии данных, а после обновления их выводы стали ошибочны

Врач или СППР чаще всего принимают решения на основе отдельных результатов исследования, а не всего протокола, так что тут похоже на работу с разными версиями ресурса `Observation`.

**Reply by Vrodvi** [reply]
Поддерживаю замечание про Task.

Вот тут описаны границы применения данного ресурса:
https://www.hl7.org/fhir/task.html#bnr
https://www.hl7.org/fhir/servicerequest.html#bnr

> A Task resource often exists in parallel with clinical resources. For example, a Task could request fulfillment of a ServiceRequest ordering a Procedure that would result in a Procedure, Observation, DiagnosticReport, ImagingStudy or similar resource.

> As stated above, the task resource tracks the state of a task, enabling systems to ensure that tasks are completed. This information is kept separate from the operational details necessary to complete the task, as those details vary across and even within workflows. That detail is expected to be carried in the referenced `focus` of the task.

Из описания следует, что Task только дополняет исходные ресурсы (указанные в атрибуте `focus`), и нужен только для отслеживания процесса выполнения, но не для описания того что должно быть выполнено.

@ai-fhir Можете пояснить почему выбраны именно ресурсы Task (а не ресурс ServiceRequest, например) и для кодирования какой информации вы их предполагаете использовать?

**Reply by ai-fhir** [reply]
Под задачей "выгрузкой результатов лаб исследований" подразумевается процесс обмена лабораторными исследованиями, который существует сейчас в формате HL7 CDA. Можно сказать, что это обмен протоколами.

Ресурс DiagnosticReport используется в протоколе, действительно, можно искать в Bundle ресурс DiagnosticReport и по его коду делать вывод про содержание документа. Почему я пишу про Task? Потому что протокол лабораторного исследования у нас сохраняет согласованность с текущим протоколом, который реализован по HL7 CDA. Текущий протокол сильно опирается на понятие "направление", которые в нашем случае выражаются связкой ServiceRequest + Task. Отсюда и появляется Task как сущность, которая задает обобщенную информацию о заказе и в том числе указывает, что в Bundle лежит заказ лабораторных услуг и результаты этого заказа. Почему Task, а не ServiceRequest главный в связке ServiceRequest + Task? Потому что ServiceRequest в заказе может быть много. 

СППР часто опираются на отдельные Observation, все верно. Но Observation должен как-то появиться в нашей системе, допустим мы запрашиваем результаты по одному заказу и они разные во времени. Какие наши действия как у принимающей системы? Должны мы обновить Observation? Старые удалить, а новые залить? Что нам делать с выводами СППР, которые были сделаны на основе результатов, которые изменились? С иммутабельными документами это достаточно понятная задача.

**Reply by IvanekOS** [reply]
Мне не очень понятно, что значит: 

> сильно опирается на понятие "направление"

В текущем руководстве по реализации СЭМД "Протокол лабораторного исследования" редакции 4 направление может как присутствовать, так и отсутствовать в документе. В будущей редакции мы действительно планируем внедрять справочник "Федеральный справочник лабораторных исследований. Справочник иерархии лабораторных профилей" (OID: 1.2.643.5.1.13.13.99.2.37) для того, чтобы врач мог назначать лабораторные исследования с исходно известным перечнем тестов и эта информация может быть отражена в направлении. Однако эта опция остается опциональной.

Аналогично мы планируем модернизировать группировочные элементы в секции "Результаты лабораторных исследований" в части поля "Указание произвольной группировки исследований" добавив справочник "Федеральный справочник лабораторных исследований. Справочник иерархии лабораторных профилей" (OID: 1.2.643.5.1.13.13.99.2.37).

Связь между блоками направление и результатом в случае их наличия будет осуществляться с помощью кодов по справочнику "Федеральный справочник лабораторных исследований. Справочник иерархии лабораторных профилей" (OID: 1.2.643.5.1.13.13.99.2.37) не более.

**Reply by IvanekOS** [reply]
Вопрос в части выбора Collection и Bundle с типом document:
Чем вызвана потребность уходить от уже разработанного и описанного варианта Bundle с типом document? Тем более, что Вы указываете:
> Переизобретаем часть логики, которая уже заложена в Bundle с типом document

Мне кажется корректнее соответствовать общепринятым ресурсам, чем изобретать что-то новое. Большой плюс Bundle с типом document - общие правила подписи и отображения документов, что помогает нам в части:

> протокол лабораторного исследования у нас сохраняет согласованность с текущим протоколом, который реализован по HL7 CDA

**Reply by Vrodvi** [reply]
Не все системы могут или хотят связываться с избыточной сложностью, накладываемой вариантом работы через "Документы", в том числе с функциональностью подписи и отображением для человека.

Преимущество FHIR как раз состоит в возможности определить наполнение данных в виде единообразного набора ресурсов, которые затем можно использовать и как документы и как сообщения, или просто как набор данных для машинной обработки (не затрагивая человекочитаемость).

В этом смысле, действительно, важно составить профили ресурсов для лабораторного протокола так, чтобы они одинаково легко могли быть использованы и как отдельный набор данных так и в виде человекочитаемого и подписанного документа.

Набор профилей, описывающих протокол лабораторного исследования - это только основа.
Над ней можно построить дополнительные профили:
- для обмена по технологии REST (bundle.type = collection)
- для обмена сообщениями (bundle.type=message, MessageHeader) 
- для обмена документами (bundle.type=document, Composition) 

Очевидно, что при реализации обмена документами понадобится гораздо больше мета-данных и сопутствующих ограничений связанных с подписью и человекочитаемостью, но это не должно влиять на структуру и наполнение самих данных протокола.
Вопрос обмена данными (rest, сообщение, документы) должен быть второстепенным вопросом на данный момент.
Если профили\ресурсы построены так что они имеют смысл только в рамках бандла определённого типа - то скорее всего что-то неправильно смоделировано.
